packages/tpetra/core/src/Tpetra_CrsMatrix_def.hpp

```diff
@@ -5442,7 +5449,14 @@ namespace Tpetra {
          std::runtime_error, "X and Y may not alias one another.");
     }
 
-    lclMatrix_->apply (X_lcl, Y_lcl, mode, alpha, beta);
+    LocalOrdinal nrows = getNodeNumRows();
+    LocalOrdinal maxRowImbalance = 0;
+    if(nrows != 0)
+      maxRowImbalance = getNodeMaxNumRowEntries() - (getNodeNumEntries() / getNodeNumRows());
```

Comment:
- I'm not sure if one of these (integer vs. proportion) is necessarily better than the other. One contrived example where absolute integer is better: all rows have 1 entry, except for one row that has 4 entries. It would be faster to use normal SPMV, but the imbalance proportion is 4x so that would probably trigger the imbalanced code path.
- I agree that the "excess of average" is not what the Behavior function name suggests. The reason I did that is to reuse what's already there from Tim's work, and that in the target use case (large, sparse matrix with a handful of dense rows) the average is much smaller than the max row length, so the difference between "excess of average" and just "long row length" is small. I realize now that Tim does TeamPolicy unpacking all the time, so I'm the only one using ``Behavior::longRowMinNumEntries()``. I think I'll change this like you suggest.
- Integer division should be OK here: IDIV has a max latency of 95 cycles, which is 3e-8s at 3 GHz. So about 0.1% of a 30us kernel launch (it's not nothing, but I would take this over precomputing it in computeLocalConstants() and then reading it each time).

Reply:
@kddevin OK, I renamed that to ``Behavior::rowImbalanceThreshold()``/``TPETRA_ROW_IMBALANCE_THRESHOLD``. The exact meaning is documented here:
```
  /// \brief Threshold for deciding if a local matrix is "imbalanced" in
  /// the number of entries per row. The threshold is compared against
  /// the difference between maximum row length and average row length.
  ///
  /// The threshold is measured in max number of entries in excess of the
  /// average (it is not a proportion between max and average).
  ///
  /// If the "imbalance" of a local matrix is greater than this threshold,
  /// a different algorithm may be used for some operations like 
  /// sparse matrix-vector multiply, packAndPrepare, and
  /// unpackAndCombine.  You may control this at run time via the
  /// <tt>TPETRA_ROW_IMBALANCE_THRESHOLD</tt> environment variable.
```

Reply:
I should have been more specific about the integer division.
I was not thinking about its performance.  Rather, I was thinking about its truncation.  From my other work, I am used to thinking of imbalance as a percentage, and we've been burnt in the past by integer division's truncation.  I can't quickly think of a case here where it would be a problem, but I thought I'd ask the question.

Reply:
@kddevin Ah, sorry. I don't think truncation will hurt us here, because it can only change the mean entries per row by either 0 or 1. So the real threshold is somewhere between ``rowImbalanceThreshold() - 1`` (maximum truncation) and ``rowImbalanceThreshold()`` (no truncation). The mean could round down to 0, but the "imbalance" will never be negative.